### PR TITLE
SCHEMACODE: Generate test data archive

### PR DESCRIPTION
**Pull request by TheChymera**

### CODE REVIEW

**Comment by TheChymera** [comment]
@effigies I'd suggest merging this on its own, and having a separate PR for github workflow integration which would also include pypi upload, and which hopefully won't but might end up being a more open-ended PR going beyond the sprint duration.

This obviously does not get executed on regular test runs, but it works, and is easy to test manually if you want to:

```console
chymera@darkhost ~/src/bids-specification/tools/schemacode/bidsschematools $ export BIDSSCHEMATOOLS_RELEASE_MODE_ACTIVATED=1
chymera@darkhost ~/src/bids-specification/tools/schemacode/bidsschematools $ pytest -vvs tests/test_make_testdata.py 
===================================== test session starts ======================================
platform linux -- Python 3.10.6, pytest-7.1.2, pluggy-1.0.0 -- /usr/bin/python3.10
cachedir: .pytest_cache
rootdir: /home/chymera/src/bids-specification/tools/schemacode, configfile: pyproject.toml
plugins: mock-3.8.2, pkgcore-0.12.13, pyfakefs-4.6.3
collected 1 item                                                                               

tests/test_make_testdata.py::test_make_archive remote: Enumerating objects: 122, done.
remote: Counting objects: 100% (122/122), done.
remote: Compressing objects: 100% (88/88), done.
remote: Total 122 (delta 30), reused 104 (delta 29), pack-reused 0
Receiving objects: 100% (122/122), 351.23 KiB | 91.00 KiB/s, done.
Resolving deltas: 100% (30/30), done.
Updating files: 100% (198/198), done.
remote: Enumerating objects: 20, done.
remote: Counting objects: 100% (20/20), done.
remote: Compressing objects: 100% (17/17), done.
remote: Total 20 (delta 1), reused 20 (delta 1), pack-reused 0
Receiving objects: 100% (20/20), 342.97 KiB | 96.00 KiB/s, done.
Resolving deltas: 100% (1/1), done.
Updating files: 100% (23/23), done.
PASSED

====================================== 1 passed in 14.65s ======================================
chymera@darkhost ~/src/bids-specification/tools/schemacode/bidsschematools $ unset BIDSSCHEMATOOLS_RELEASE_MODE_ACTIVATED
chymera@darkhost ~/src/bids-specification/tools/schemacode/bidsschematools $ pytest -vvs tests/test_make_testdata.py 
===================================== test session starts ======================================
platform linux -- Python 3.10.6, pytest-7.1.2, pluggy-1.0.0 -- /usr/bin/python3.10
cachedir: .pytest_cache
rootdir: /home/chymera/src/bids-specification/tools/schemacode, configfile: pyproject.toml
plugins: mock-3.8.2, pkgcore-0.12.13, pyfakefs-4.6.3
collected 1 item                                                                               

tests/test_make_testdata.py::test_make_archive SKIPPED (Release mode is not activated. Not generating static testdata archive. In order to force testdata archive generation run: `export BIDSSCHEMATOOLS_RELEASE_MODE_ACTIVATED=1`.)

====================================== 1 skipped in 0.01s ======================================
```

File: tools/schemacode/bidsschematools/tests/test_make_testdata.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    env = os.environ.get(var)
+    return pytest.mark.skipif(
+        not env,
+        reason="Release mode is not activated. Not generating static testdata archive. "
+        "In order to force testdata archive generation run: "
+        "`export BIDSSCHEMATOOLS_RELEASE_MODE_ACTIVATED=1`.",
+    )
+
+
+@require_env("BIDSSCHEMATOOLS_RELEASE_MODE_ACTIVATED")
```
Comment (effigies) [inline tools/schemacode/bidsschematools/tests/test_make_testdata.py:21]: I'd try to write this decorator a bit more generically, and then add a reason where called. And I'd also suggest a shorter environment variable:

```suggestion
def require_env(var, reason=None):
    instructions = f"To activate this test/feature `export {var}=1` and re-run."
    return pytest.mark.skipif(
        not os.getenv(var),
        reason=f"{reason} {instructions}" if reason else instructions
    )


@require_env("BST_RELEASE", "Release mode inactive; not generating test data archive.")
```

File: tools/schemacode/bidsschematools/tests/test_make_testdata.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    bst_version = open(version_file).read().rstrip("\n")
+    archive_name = f"bidsschematools-testdata-{bst_version}"
+    archive_path = f"/tmp/{archive_name}.tar.gz"
+
+    with tempfile.TemporaryDirectory() as temp_path:
+        shutil.copytree(bids_examples, os.path.join(temp_path, "bids-examples"))
+        shutil.copytree(bids_error_examples, os.path.join(temp_path, "bids-error-examples"))
+
+        with tarfile.open(archive_path, "w:gz") as tar:
+            tar.add(temp_path, arcname=archive_name)
```
Comment (effigies) [inline tools/schemacode/bidsschematools/tests/test_make_testdata.py:47]: I think we can avoid making an extra copy:

```suggestion
    with tarfile.open(archive_path, "w:gz") as tar:
        tar.add(bids_examples, arcname=f"{archive_name}/bids-examples")
        tar.add(bids_error_examples, arcname=f"{archive_name}/bids-error-examples")
```

**Comment by TheChymera** [comment]
@effigies fixed --- I think it's better to keep the envvar name a bit more explicit than what you suggested. You never know what people have dangling in their env.

**Comment by bendhouseart** [comment]
Should the archive paths be stored as environment variables instead of hardcoded paths? ;)

**Review by effigies — APPROVED** [review]

**Review by bendhouseart — APPROVED** [review]

**Pull request merged**